src/opensles/AudioStreamOpenSLES.cpp

```diff
@@ -123,35 +123,61 @@ SLresult AudioStreamOpenSLES::finishCommonOpen(SLAndroidConfigurationItf configI
-                (kHighLatencyBufferSizeMillis * sampleRate) / kMillisPerSecond;
+int32_t AudioStreamOpenSLES::calculateOptimalBufferQueueLength() {
+    int32_t queueLength = kBufferQueueLengthDefault;
+    int32_t likelyFramesPerBurst = estimateNativeFramesPerBurst();
+    // What is required based on capacity?
+    if (mBufferCapacityInFrames > 0) {
+        int32_t queueLengthFromCapacity = (mBufferCapacityInFrames + likelyFramesPerBurst - 1)
+                / likelyFramesPerBurst;
+        queueLength = std::max(queueLength, queueLengthFromCapacity);
+    }
+    // What is required based on requested callback size?
+    if (mFramesPerCallback > 0) {
+        int32_t minCapacity = 2 * mFramesPerCallback;
+        int32_t queueLengthFromCallbackSize = (minCapacity + likelyFramesPerBurst - 1)
+                                  / likelyFramesPerBurst;
+        queueLength = std::max(queueLength, queueLengthFromCallbackSize);
+    }
+    queueLength = std::min(queueLength, kBufferQueueLengthMax);
+    // TODO Investigate the effect of queueLength on latency for normal streams. (not low latency)
+    return queueLength;
+}
 
-        // For high latency streams, use a larger buffer size.
-        // Performance Mode support was added in N_MR1 (7.1)
-        if (getSdkVersion() >= __ANDROID_API_N_MR1__
+/**
+ * The best information we have is if DefaultStreamValues::FramesPerBurst
+ * was set by the app based on AudioManager.PROPERTY_OUTPUT_FRAMES_PER_BUFFER.
+ * Without that we just have to guess.
+ * @return
+ */
+int32_t AudioStreamOpenSLES::estimateNativeFramesPerBurst() {
+    int32_t framesPerBurst = DefaultStreamValues::FramesPerBurst;
+    framesPerBurst = std::max(framesPerBurst, 16);
+    // Calculate the size of a fixed duration high latency buffer based on sample rate.
+    int32_t sampleRate = (mSampleRate > 0) ? mSampleRate : 48000;
+    int32_t framesPerHighLatencyBuffer =
+            (kHighLatencyBufferSizeMillis * sampleRate) / kMillisPerSecond;
+    // For high latency streams, use a larger buffer size.
+    // Performance Mode support was added in N_MR1 (7.1)
+    if (getSdkVersion() >= __ANDROID_API_N_MR1__
             && mPerformanceMode != PerformanceMode::LowLatency
-            && mFramesPerBurst < framesPerHighLatencyBuffer) {
-            // Find a multiple of framesPerBurst >= framesPerHighLatencyBuffer.
-            int32_t numBursts = (framesPerHighLatencyBuffer + mFramesPerBurst - 1) / mFramesPerBurst;
-            mFramesPerBurst *= numBursts;
-            LOGD("AudioStreamOpenSLES:%s() NOT low latency, set mFramesPerBurst = %d",
-                 __func__, mFramesPerBurst);
-        }
-        mFramesPerCallback = mFramesPerBurst;
-    }
+            && framesPerBurst < framesPerHighLatencyBuffer) {
+        // Find a multiple of framesPerBurst >= framesPerHighLatencyBuffer.
+        int32_t numBursts = (framesPerHighLatencyBuffer + framesPerBurst - 1) / framesPerBurst;
```

Comment:
Seems like we are using this method for finding a multiple in the code in quite a few places. Can we write a short helper function for this so the code is easier to read?

Reply:
Done